src/psyclone/tests/psyir/frontend/fparser2_subroutine_handler_test.py

```diff
@@ -93,3 +96,124 @@ def test_subroutine_handler(parser, code, expected):
+        "    integer :: my_func\n"
+        "    my_func = 1\n"
+        "  end function my_func\n"
+        "end module\n")
+    expected = (
+        "module a\n"
+        "  implicit none\n\n"
+        "  public :: my_func\n\n"
+        "  contains\n"
+        "  function my_func()\n"
+        "    integer :: my_func\n"
+        "\n"
+        "    my_func = 1\n"
+        "\n"
+        "  end function my_func\n"
+        "\n"
+        "end module a\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    # Check PSyIR nodes are being created
+    assert isinstance(psyir, Container)
+    routines = psyir.walk(Routine)
+    assert len(routines) == 1
+    assert isinstance(routines[0].return_symbol, DataSymbol)
+    assert (routines[0].return_symbol.datatype.intrinsic ==
+            ScalarType.Intrinsic.INTEGER)
+    assert psyir.parent is None
+    writer = FortranWriter()
+    result = writer(psyir)
+    assert result == expected
+
+
+FN1_IN = ("  function my_func() result(my_val)\n"
+          "    real :: my_val\n"
+          "    my_val = 1.0\n"
+          "  end function my_func\n")
+FN2_IN = ("  real function my_func() result(my_val)\n"
+          "    my_val = 1.0\n"
+          "  end function my_func\n")
+FN3_IN = ("  real(wp) function my_func() result(my_val)\n"
+          "    my_val = 1.0\n"
+          "  end function my_func\n")
+EXPECTED_FN_OUT = ("  function my_func() result(my_val)\n"
+                   "    real :: my_val\n\n"
+                   "    my_val = 1.0\n\n"
+                   "  end function my_func\n")
+EXPECTED_FN3_OUT = ("  function my_func() result(my_val)\n"
+                    "    real(kind=wp) :: my_val\n\n"
+                    "    my_val = 1.0\n\n"
+                    "  end function my_func\n")
+
+
+@pytest.mark.parametrize("code,expected",
+                         [(FN1_IN, EXPECTED_FN_OUT),
+                          (FN2_IN, EXPECTED_FN_OUT),
+                          (FN3_IN, EXPECTED_FN3_OUT)])
+def test_function_result_suffix(fortran_reader, code, expected):
+    '''
+    Test that we handle a Fortran function with the return value specified
+    using the 'result()' suffix. We test when the type is specified by a
+    declaration inside the function or by a type specifier in the function
+    prefix.
+
+    '''
+    code = (
+        "module a\n"
+        "use kind_params, only: wp\n"
+        "contains\n"
+        "{0}end module\n".format(code))
+    psyir = fortran_reader.psyir_from_source(code)
+    # Check PSyIR nodes are being created
+    assert isinstance(psyir, Container)
+    routines = psyir.walk(Routine)
+    assert len(routines) == 1
+    assert (routines[0].return_symbol is
+            routines[0].symbol_table.lookup("my_val"))
+    writer = FortranWriter()
+    result = writer(psyir)
+    assert expected in result
+
+
+def test_function_missing_return_type(fortran_reader):
+    '''
+    Test that we reject a Fortran function if there is no explicit declaration
```

Comment:
Is "reject" the right word? Maybe "turn into a CodeBlock?

Are we creating a codeblock instead of failing because my_func can have a type with the implicit types Fortran rules? It may be worth adding a comment explaining it.

Reply:
Reworded and expanded.